README.md
<div align="center">
  
# vercel-pdf-converter

[![GitHub](https://img.shields.io/github/license/mashape/apistatus.svg)](https://github.com/BetaHuhn/vercel-pdf-converter/blob/master/LICENSE) ![Version](https://img.shields.io/github/package-json/v/betahuhn/vercel-pdf-converter) ![Dependencies](https://img.shields.io/david/betahuhn/vercel-pdf-converter)

Vercel function which generates PDFs from Webpages.

</div>

## 👋 Introduction

This repo contains the code for a simple [Vercel](https://vercel.com) function which generates a PDF file from any webpage using a [headless Chrome instance](https://github.com/puppeteer/puppeteer).

## 🚀 Get started

[![Deploy with Vercel](https://vercel.com/button)](https://vercel.com/new/git/external?repository-url=https%3A%2F%2Fgithub.com%2FBetaHuhn%2Fvercel-pdf-converter)

### Or setup manually:

Clone this repository:

```sh
git clone https://github.com/BetaHuhn/vercel-pdf-converter
```

Install all dependencies:

```sh
npm install
```

Login to your Vercel account and setup a project:

```sh
vercel
```

Run the function locally:

```sh
npm run develop
```

Deploy to Vercel in production:

```sh
npm run deploy 
```

## 📚 Usage

After you've deployed the function, you can use it by placing your function's domain infront of any URL:

```
to-pdf.vercel.app/https://github.com/BetaHuhn/vercel-pdf-converter
```

The function will then generate a PDF of that URL and return it as a downloadable file.

Here is how this GitHub Page looks as a PDF:

![example](https://cdn.mxis.ch/assets/vercel-pdf-converter/preview.png)

## 💻 Development

Issues and PRs are very welcome!

This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html). To see differences with previous versions refer to the [CHANGELOG](CHANGELOG.md).

## ❔ About

This project was developed by me ([@betahuhn](https://github.com/BetaHuhn)) in my free time. If you want to support me:

[![Donate via PayPal](https://img.shields.io/badge/paypal-donate-009cde.svg)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=394RTSBEEEFEE)

## License

Copyright 2020 Maximilian Schiller

This project is licensed under the MIT License - see the [LICENSE](LICENSE) file for details.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [v1.0.0] - 2020-11-29

### Added
- everything


api/index.js
const { getPdf } = require('../service/convert')

// Cache header max age
const maxAge = 24 * 60 * 60

module.exports = async (req, res) => {
	try {
		// Only allow GET requests
		if (req.method !== 'GET') return res.status(405).end()

		// Strip leading slash from request path
		const url = req.url.replace(/^\/+/, '')

		// Block favicon.ico requests from reaching puppeteer
		if (url === 'favicon.ico') return res.status(404).end()

		console.log(`Converting: ${ url }`)
		const pdfBuffer = await getPdf(url)

		if (!pdfBuffer) return res.status(400).send('Error: could not generate PDF')

		// Instruct browser to cache PDF for maxAge ms
		if (process.env.NODE_ENV !== 'development') res.setHeader('Cache-control', `public, max-age=${ maxAge }`)

		// Set Content type to PDF and send the PDF to the client
		res.setHeader('Content-type', 'application/pdf')
		res.send(pdfBuffer)

	} catch (err) {
		if (err.message === 'Protocol error (Page.navigate): Cannot navigate to invalid URL')
			return res.status(404).end()

		console.error(err)
		res.status(500).send('Error: Please try again.')
	}
}

service/convert.js
import chrome from 'chrome-aws-lambda'
import { addExtra } from 'puppeteer-extra'
import AdblockerPlugin from 'puppeteer-extra-plugin-adblocker'

// Workaround, see https://github.com/berstend/puppeteer-extra/issues/93#issuecomment-712364816
import StealthPlugin from 'puppeteer-extra-plugin-stealth'
import ChromeAppPlugin from 'puppeteer-extra-plugin-stealth/evasions/chrome.app'
import ChromeCsiPlugin from 'puppeteer-extra-plugin-stealth/evasions/chrome.csi'
import ChromeLoadTimes from 'puppeteer-extra-plugin-stealth/evasions/chrome.loadTimes'
import ChromeRuntimePlugin from 'puppeteer-extra-plugin-stealth/evasions/chrome.runtime'
import IFrameContentWindowPlugin from 'puppeteer-extra-plugin-stealth/evasions/iframe.contentWindow'
import MediaCodecsPlugin from 'puppeteer-extra-plugin-stealth/evasions/media.codecs'
import NavigatorLanguagesPlugin from 'puppeteer-extra-plugin-stealth/evasions/navigator.languages'
import NavigatorPermissionsPlugin from 'puppeteer-extra-plugin-stealth/evasions/navigator.permissions'
import NavigatorPlugins from 'puppeteer-extra-plugin-stealth/evasions/navigator.plugins'
import NavigatorVendor from 'puppeteer-extra-plugin-stealth/evasions/navigator.vendor'
import NavigatorWebdriver from 'puppeteer-extra-plugin-stealth/evasions/navigator.webdriver'
import SourceUrlPlugin from 'puppeteer-extra-plugin-stealth/evasions/sourceurl'
import UserAgentOverridePlugin from 'puppeteer-extra-plugin-stealth/evasions/user-agent-override'
import WebglVendorPlugin from 'puppeteer-extra-plugin-stealth/evasions/webgl.vendor'
import WindowOuterDimensionsPlugin from 'puppeteer-extra-plugin-stealth/evasions/window.outerdimensions'

// Configure puppeteer-extra plugins
const puppeteer = addExtra(chrome.puppeteer)
const plugins = [
	AdblockerPlugin({ blockTrackers: true }),
	StealthPlugin(),
	ChromeAppPlugin(),
	ChromeCsiPlugin(),
	ChromeLoadTimes(),
	ChromeRuntimePlugin(),
	IFrameContentWindowPlugin(),
	MediaCodecsPlugin(),
	NavigatorLanguagesPlugin(),
	NavigatorPermissionsPlugin(),
	NavigatorPlugins(),
	NavigatorVendor(),
	NavigatorWebdriver(),
	SourceUrlPlugin(),
	UserAgentOverridePlugin(),
	WebglVendorPlugin(),
	WindowOuterDimensionsPlugin()
]

// Or just use puppeteer directly
// import puppeteer from 'puppeteer-core'

const isDev = process.env.NODE_ENV === 'development'

// Path to chrome executable on different platforms
const chromeExecutables = {
	linux: '/usr/bin/chromium-browser',
	win32: 'C:\\Program Files (x86)\\Google\\Chrome\\Application\\chrome.exe',
	darwin: '/Applications/Google Chrome.app/Contents/MacOS/Google Chrome'
}

export const getOptions = async (isDev) => {

	// During development use local chrome executable
	if (isDev) {
		return {
			args: [],
			executablePath: chromeExecutables[process.platform] || chromeExecutables.linux,
			headless: true
		}
	}

	// Else, use the path of chrome-aws-lambda and its args
	return {
		args: chrome.args,
		executablePath: await chrome.executablePath,
		headless: chrome.headless
	}
}

export const getPdf = async (url) => {

	// Start headless chrome instance
	const options = await getOptions(isDev)
	const browser = await puppeteer.launch(options)

	// Load all plugins manually
	for (const plugin of plugins) {
		await plugin.onBrowser(browser)
	}

	const page = await browser.newPage()

	// Visit URL and wait until everything is loaded (available events: load, domcontentloaded, networkidle0, networkidle2)
	await page.goto(url, { waitUntil: 'networkidle2', timeout: 8000 })

	// Scroll to bottom of page to force loading of lazy loaded images
	await page.evaluate(async () => {
		await new Promise((resolve) => {
			let totalHeight = 0
			const distance = 100
			const timer = setInterval(() => {
				const scrollHeight = document.body.scrollHeight
				window.scrollBy(0, distance)
				totalHeight += distance

				if (totalHeight >= scrollHeight) {
					clearInterval(timer)
					resolve()
				}
			}, 5)
		})
	})

	// Tell Chrome to generate the PDF
	await page.emulateMediaType('screen')
	const buffer = await page.pdf({
		format: 'A4',
		displayHeaderFooter: true,
		headerTemplate: '',
		footerTemplate: '',
		printBackground: true
	})

	// Close chrome instance
	await browser.close()

	return buffer
}

public/main.js
/* eslint-disable no-unused-vars */
const go = document.getElementById('go')
const host = document.getElementById('host')
const input = document.getElementById('url')
const error = document.getElementById('error')
const errorMsg = document.getElementById('error-msg')

if (host) host.innerText = `${ window.location.host }/`

const errors = {
	400: 'Error: Could not generate PDF.',
	404: 'Error: Invalid URL.',
	500: 'Error: Please try again.'
}

const status = (location.search.split('status=')[1] || '').split('&')[0]
if (status.length > 0) {
	displayError(errors[status])
}

function generate() {
	if (go.href === undefined || go.href.length < 1) return displayError('Please fill out the form.')
	go.innerText = 'Generating...'
	window.location.href = go.href
}

input.addEventListener('keyup', function(event) {
	if (event.key === 'Enter' || event.keyCode === 13) {
		event.preventDefault()
		generate()
	}
})

function onInput(input) {
	go.href = `http://${ window.location.host }/${ input.value }`
}

function copy() {
	const value = `${ window.location.host }/`
	const tempInput = document.createElement('INPUT')
	document.body.appendChild(tempInput)
	tempInput.setAttribute('value', value)
	tempInput.select()
	document.execCommand('copy')
	document.body.removeChild(tempInput)
	host.title = 'Copied!'
	setTimeout(function() {
		host.title = 'Click to copy'
	}, 1000)
}

function displayError(msg) {
	errorMsg.innerText = msg || 'An unkown error ocurred.'
	error.style.display = 'block'
}